src/components/timeline/TrackVisualProgressGraph.js

```diff
@@ -0,0 +1,346 @@
+      let x = 0;
+      let y = 0;
+      let yOld = 0;
+      for (let i = 0; i < graphData.length; i++) {
+        // Create a path for the top of the chart. This is the line that will have
+        // a stroke applied to it.
+        x = (deviceWidth * (graphData[i].timestamp - rangeStart)) / rangeLength;
+        // Add on half the stroke's line width so that it won't be cut off the edge
+        // of the graph.
+        const unitGraphCount = graphData[i].percent / 100;
+        y =
+          innerDeviceHeight -
+          innerDeviceHeight * unitGraphCount +
+          deviceLineHalfWidth;
+        if (i === 0) {
+          // This is the first iteration, only move the line.
+          ctx.moveTo(x, y);
+        } else {
+          // Plot an extra point to generate a step-wise graph
+          ctx.lineTo(x, yOld);
+          ctx.lineTo(x, y);
+        }
+        yOld = y;
+      }
+      // The samples range ends at the time of the last sample, plus the interval.
+      // Draw this last bit.
+      x = (deviceWidth * (rangeEnd - rangeStart)) / rangeLength;
+      ctx.lineTo(x, y);
+
+      // Don't do the fill yet, just stroke the top line.
+      ctx.stroke();
+
+      // After doing the stroke, continue the path to complete the fill to the bottom
+      // of the canvas.
+      ctx.lineTo(x + interval, deviceHeight);
+      ctx.lineTo(
+        (deviceWidth * (graphData[0].timestamp - rangeStart)) / rangeLength +
+          interval,
+        deviceHeight
+      );
+      ctx.fill();
+    }
+  }
+
+  _scheduleDraw() {
+    if (!this._requestedAnimationFrame) {
+      this._requestedAnimationFrame = true;
+      window.requestAnimationFrame(() => {
+        this._requestedAnimationFrame = false;
+        const canvas = this._canvas;
+        if (canvas) {
+          this.drawCanvas(canvas);
+        }
+      });
+    }
+  }
+
+  _takeCanvasRef = (canvas: HTMLCanvasElement | null) => {
+    this._canvas = canvas;
+  };
+
+  render() {
+    this._scheduleDraw();
+
+    return (
+      <canvas
+        className="timelineTrackVisualProgressCanvas"
+        ref={this._takeCanvasRef}
+      />
+    );
+  }
+}
+
+type OwnProps = {|
+  +graphData: VisualProgress[],
+  +lineWidth: CssPixels,
+  +graphHeight: CssPixels,
+  +graphStrokeColor: string,
+  +graphFillColor: string,
+  +graphDotColor: string,
+  +graphDotTooltipText: string,
+|};
+
+type StateProps = {|
+  +rangeStart: Milliseconds,
+  +rangeEnd: Milliseconds,
+  +interval: Milliseconds,
+|};
+
+type DispatchProps = {||};
+
+type Props = {|
+  ...SizeProps,
+  ...ConnectedProps<OwnProps, StateProps, DispatchProps>,
+|};
+
+type State = {|
+  hoveredCounter: null | number,
+  mouseX: CssPixels,
+  mouseY: CssPixels,
+|};
+
+/**
+ * The visual progress track graph takes visual progress information from visual metrics, and renders it as a
+ * graph in the timeline.
+ */
+class TrackVisualProgressGraphImpl extends React.PureComponent<Props, State> {
```

Comment:
I'm finding the combination of names `VisualProgress`, `VisualMetrics` and `graphData` to be a bit confusing. I generally find that having clear terminology helps the maintainability of the code.

It seems like `VisualMetrics` is all the data, and then there is a `VisualProgress` property, but the shape of the data is also called `VisualProgress`. The component is called `TrackVisualProgressGraph`. There is a `Prop` called `graphData`, but it's type is called `VisualProgress`.

Do you have any ideas here on how to make this more clear?

Reply:
Hi Greg, I'm thinking of renaming the shape and type of the data from `VisualProgress` to `ProgressGraphData`, and the `Prop` from `graphData` to `progressGraphData`. I think majority of the confusion comes from VisualProgress being a property and also a type.

Reply:
Thanks for thinking this through!